core/node/libp2p.go

```diff
@@ -0,0 +1,578 @@
+	if cfg.Swarm.DisableRelay {
+		// Enabled by default.
+		opts.Opts = append(opts.Opts, libp2p.DisableRelay())
+	} else {
+		relayOpts := []relay.RelayOpt{relay.OptDiscovery}
+		if cfg.Swarm.EnableRelayHop {
+			relayOpts = append(relayOpts, relay.OptHop)
+		}
+		opts.Opts = append(opts.Opts, libp2p.EnableRelay(relayOpts...))
+	}
+	return
+}
+
+func P2PAutoRealy(cfg *config.Config) (opts Libp2pOpts, err error) {
+	// enable autorelay
+	if cfg.Swarm.EnableAutoRelay {
+		opts.Opts = append(opts.Opts, libp2p.EnableAutoRelay())
+	}
+	return
+}
+
+func P2PDefaultTransports() (opts Libp2pOpts, err error) {
+	opts.Opts = append(opts.Opts, libp2p.DefaultTransports)
+	return
+}
+
+func P2PQUIC(cfg *config.Config) (opts Libp2pOpts, err error) {
+	if cfg.Experimental.QUIC {
+		opts.Opts = append(opts.Opts, libp2p.Transport(libp2pquic.NewTransport))
+	}
+	return
+}
+
+func P2PNoSecurity() (opts Libp2pOpts) {
+	opts.Opts = append(opts.Opts, libp2p.NoSecurity)
+	// TODO: shouldn't this be Errorf to guarantee visibility?
+	log.Warningf(`Your IPFS node has been configured to run WITHOUT ENCRYPTED CONNECTIONS.
+		You will not be able to connect to any nodes configured to use encrypted connections`)
+	return opts
+}
+
+type P2PHostIn struct {
+	fx.In
+
+	Repo          repo.Repo
+	Validator     record.Validator
+	HostOption    HostOption
+	RoutingOption RoutingOption
+	ID            peer.ID
+	Peerstore     peerstore.Peerstore
+
+	Opts [][]libp2p.Option `group:"libp2p"`
+}
+
+type BaseRouting routing.IpfsRouting
+type P2PHostOut struct {
+	fx.Out
+
+	Host    host.Host
+	Routing BaseRouting
+}
+
+func P2PHost(mctx MetricsCtx, lc fx.Lifecycle, params P2PHostIn) (out P2PHostOut, err error) {
+	opts := []libp2p.Option{libp2p.NoListenAddrs}
+	for _, o := range params.Opts {
+		opts = append(opts, o...)
+	}
+
+	ctx := lifecycleCtx(mctx, lc)
+
+	opts = append(opts, libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
+		r, err := params.RoutingOption(ctx, h, params.Repo.Datastore(), params.Validator)
+		out.Routing = r
+		return r, err
+	}))
+
+	out.Host, err = params.HostOption(ctx, params.ID, params.Peerstore, opts...)
+	if err != nil {
+		return P2PHostOut{}, err
+	}
+
+	// this code is necessary just for tests: mock network constructions
+	// ignore the libp2p constructor options that actually construct the routing!
+	if out.Routing == nil {
+		r, err := params.RoutingOption(ctx, out.Host, params.Repo.Datastore(), params.Validator)
+		if err != nil {
+			return P2PHostOut{}, err
+		}
+		out.Routing = r
+		out.Host = routedhost.Wrap(out.Host, out.Routing)
+	}
+
+	lc.Append(fx.Hook{
+		OnStop: func(ctx context.Context) error {
+			return out.Host.Close()
+		},
+	})
+
+	return out, err
+}
+
+type Router struct {
+	routing.IpfsRouting
+
+	Priority int // less = more important
+}
+
+type p2pRouterOut struct {
+	fx.Out
+
+	Router Router `group:"routers"`
+}
+
+func P2PBaseRouting(lc fx.Lifecycle, in BaseRouting) (out p2pRouterOut, dr *dht.IpfsDHT) {
```

Comment:
exported func P2PBaseRouting returns unexported type node.p2pRouterOut, which can be annoying to use